### DISCUSSION TITLE
Calling Magick from  a C++ program

### DISCUSSION THREAD

**Discussion by DocDJ** [ImageMagick/ImageMagick#6112] [Help]
### ImageMagick version

7.0.10-61 Q16 X64

### Operating system, version and so on

Windows 11 Pro

### Description

I have written an MFC dialog that is trying to call Magick using the CreateProcess function. I have replicated the parameter string I used for calling Magick from my Windows batch file version, but I am getting no output. I have a long parameter string with blanks inserted between options and values. When the program runs, a window flashes on the screen, but it disappears to quickly to see if anything is in it. I have redirected STDOUT to a file (which works for other programs in the dialog), but there is nothing in the file from Magick. Are there any caveats for what I am doing? Any tools/tips for finding out what I am doing wrong?
I have used MessageBox to display the parameter string and it matches what I did in my batch file.

**Comment by snibgo** [comment]
Without seeing your code and values of any variables, remote debugging is difficult.

What value is returned from CreateProcess()?

What values are you passing to CreateProcess()? Have you read documentation about CreateProcess(), and checked you are calling it correctly?

> I have used MessageBox to display the parameter string and it matches what I did in my batch file.

As you should know, matching what you did in a BAT file is likely wrong, eg don't double `%`, can't use search path, need full program name.

**Reply by DocDJ** [reply]
CreateProcess returns 0. Here is my parameter string (all variables are strings, div is a "/", NL is a '\n', space is a blank). All variables are CStrings which CreateProcess converts to char before they are passed).  This mechanism works for the other programs I created using CreateProcess., which were originally cmdline programs called form Batch files.
CString("mogrify -format jpg -font Rockwell -weight ExtraBold -fill gold ") +
				CString("-pointsize") + space + s_pointsize + space +
				CString("-stroke black -strokewidth ") + s_strokewidth + space + CString("-annotate ") + s_annotate_disp + space +
				CString(metadata._Desc) + NL +
				CString(metadata._Sub) + NL +
				CString(metadata._City) + div + CString(metadata._State) + div + CString(metadata._Country) + space +
				CString("-stroke none -annotate ") + s_annotate_disp + space +
				CString(metadata._Desc) + NL +
				CString(metadata._Sub) + NL +
				CString(metadata._State) + div + 
				CString(metadata._Country) + space +
				CString("-path ") + CString(outpath) + space + CString(injpg);

**Comment by snibgo** [comment]
You are using IM v7. "mogrify" uses the old v6 syntax. I suggest using "magick mogrify" instead. But this probably won't fix your problem.

What string are you passing to CreateProcess? The actual string, not an expression with variables.

I suggest you test your program with a simpler string. For example, first create an image file by running "`magick rose: r.png`" from the command line. Then test your program by passing "`magick mogrify -format jpg r.png`" to CreateProcess().

If that works, then you know you can successfully run `magick mogrify`, and the problem is in the parameters to that command.

**Comment by DocDJ** [comment]
thanks for the tip. I'll try it tomorrow.

**Comment by DocDJ** [comment]
Magick was able to run from CreateProcess and produced r.jpg as expected. So there must be some problem inside my parameter string. The question remains: where is the STDOUT from Magick? With my parameter string, Magick gives a return code of 1. The process that runs before Magick outputs to my redirected STDOUT, but there is nothing there from Magick to indicate what it did. STDERR is redirected to the same file. I added the option -debug \"all\"  which had no visible effect. I still get that window flashing before my dialog exits.

**Comment by snibgo** [comment]
> So there must be some problem inside my parameter string.

Yes.

> The question remains: where is the STDOUT from Magick?

magick will write nothing to STDOUT. If it has errors (which I expect it does in your case), it will write text to STDERR. If you haven't redirected that, I expect it goes to your console that immediately vanishes. But you say "STDERR is redirected to the same file." Perhaps there is a problem in this redirection.

I encourage you to printf or cout your command string. The problem could be trivial, like using the character '\n' instead of the two-character string "\n".

In the longer term, you probably want to trap errors, and log commands that raised errors.

**Reply by DocDJ** [reply]
Here's the result in cout:
MOGRIFY parmstring= mogrify   -format jpg -font "Rockwell" -weight ExtraBold -fill gold -pointsize 62  -stroke black -strokewidth 15 -annotate +1374 "The Radisson SAS Scandinavian lobby\n_\nCopenhagen/_/Denmark"  -stroke none -annotate +1374  "The Radisson SAS Scandinavian lobby\n_\n_/Denmark"   -path "R:\Frame_images\OUT\Baltic\1-Copen-1-3" DSC_0302.jpg

Also tried where the output path ended with a \
Nothing appears in the stderr file (now a separate file from stdout), so I still don't know what Magick doesn't like in my parameters.

The string above lets Magick exit with a return code of 0, but the output file has not been created.

**Comment by snibgo** [comment]
I can't see an obvious problem there. But maybe DSC_0302.jpg doesn't exist, or your output path doesn't exist, etc.

We know that a simplified mogrify works. So I suggest you simplify the long command until it also works.

**Reply by DocDJ** [reply]
thanks. I'll try shortening until it works. Since it DOES return 0, it doesn't seem to be troubled by the path. I tried it with the old output file deleted beforehand an no new file was created, yet still got retcode of 0. And still nothing in stderr.

**Comment by DocDJ** [comment]
I cannot get any stderr or file output. I have added -debug "all" and still get the "flashing" window. I reduced the parameter string to just 
**infile -format jpg -debug "all" outfile**  (where the file info is their full path with backslashes.) STDERR is now a completely separate file, opened with the same options as STDOUT. I'm mystified. Any hints or tricks to try?

**Comment by snibgo** [comment]
> infile -format jpg -debug "all" outfile

Without `magick`, do you expect that to work?

> I cannot get any stderr or file output.

Again, without seeing your code, I can't diagnose the problem.

Try another command that doesn't use ImageMagick, such as "`xcopy nosuch.bin out.bin`" where you *don't* have a file named `nosuch.bin`. At the CLI, this will write an error message to stderr. If your program is correct, you will get a message in your log file.

If you don't get a message in your log file, then you haven't set up CreateProcess() correctly.

**Reply by DocDJ** [reply]
Thanks for the suggestion. I tried it and you're right: I get no output. So either my invocation of the program is wrong or my re-direction is not working for this program. The current program (Magick or xcopy) has the problem, but the original program works fine (exiftool) both called from my MFC dialog. I'll have to keep testing. Thanks again!!!

**Comment by DocDJ** [comment]
After struggling with redirecting stdout and stderr, and fixing my parameter strings, I finally got Magick to run. It gave me a return code of 4. Is there a place where I can look this up?

**Comment by snibgo** [comment]
I've never seen a return code of 4 from `magick`. As far as I know, it always returns 0 or 1.

What was your command? What was the error message?

When you run that command from a console, do you get the same error message and return code?

If they are different, then how does your program get the return code? Perhaps that part of your program has a bug.

**Comment by DocDJ** [comment]
I call it from an MFC Dialogue on my PC. I do not own a console. There was no error message. in STDERR or STDOUT. I have called Magick with the same parameters from a Windows batch file with no problems. I suspect it has something to do with the redirection of STDIO and STDERR which works for the exiftool program I call before calling Magick  (from the same program). There are 4 redirected files: 2 for exiftool and 2 for Magick. Both programs are called via CreateProcess.

**Comment by snibgo** [comment]
"Console" is a name for a window that accepts text commands and displays text output.

When you run your `magick` command from the command line, what is the return code and error message?

**Comment by DocDJ** [comment]
After much reading of materials recommended above, I have re-written my CreateProcess code for redirecting both STDERR and STDOUT. Imagemagick is now getting called and it is outputting the parameter list from my program (because I added debug "all" to the parameters, but the resulting file has NONE of my mogrify actions added. Here is what Imagemagick says it is getting: ` Command line: mogrify {R:\Frame_images\IN\Baltic\1-Copen-1-3\DSC_0302.jpg} {-format} {jpg} {-debug} {all} **{-font} {Rockwell}** {-weight} {ExtraBold} {-fill} {gold} {-pointsize} {62} {-stroke} {black} {-strokewidth} {15} {-annotate} {+1374} {The Radisson SAS Scandinavian lobby\n_\nCopenhagen/_/Denmark} {-stroke} {none} {-annotate} {+1374} {The Radisson SAS Scandinavian lobby\n_\n_/Denmark} {-path} {R:\Frame_images\OUT\Baltic\1-Copen-1-3} {DSC_0302.jpg}`
I expect to get 3 lines of text in the bottom left corner. When I use this string from my batch file, it works as expected. I think there is something about the parameters such as: 
-font "Rockwell" 
being split (as shown by the { } above) by the CreateProcess code. So I suspect I need to do something to keep the pieces together to get {-font "Rockwell"}. The repeated lines are recommended in the Imagemagick docs to keep the text sharp. I already tried replaceng the " with \" which did not work.

Any hints?

**Comment by snibgo** [comment]
> So I suspect I need to do something to keep the pieces together to get {-font "Rockwell"}. 

No, they should be two separate words, as they are.

See the mogrify documentation at https://imagemagick.org/script/mogrify.php . You have a filename immediately after "mogrify". Why?

**Comment by DocDJ** [comment]
Thanks for catching that. It was supposed to be stripped before passing the parameters to IM by the calling program. I removed it and the output file still does not get any changes.

**Comment by snibgo** [comment]
> I expect to get 3 lines of text in the bottom left corner.

You haven't asked for that. You have asked for text in the top-left corner, rotated by 1374 degrees.

> When I use this string from my batch file, it works as expected. 

Does it? It shouldn't.

See `-annotate` documentation at https://imagemagick.org/script/command-line-options.php#annotate

That shows four possible formats. You have used the first format, with only one number, so that is the rotation in degrees. If you want to supply x and y offsets, use + or - before each one, eg `+0+1374` to get an x-offset of zero, and y-offset of 1374.

You could use `-gravity southWest` and zero x and y offsets to get the text in the bottom-left corner.

**Comment by DocDJ** [comment]
Thank you all so much for your time and great advice!! I added the +0 and it is behaving exactly like my batch-file version. But now that I  have a windowing dialog for the folders, I can process one or dozens of images with one operation. PS. I verified that my batch version did not use the +0 before the vertical value.
